### ISSUE TITLE
No Access to Server from VM with Virtual Network Interface: Direct attachment

### ISSUE THREAD

**Issue by err0r2k** [cockpit-project/cockpit#17740]
### Explain what happens
Updated Description:

What is the problem: I can‘t access the Server (ping, open network share, open webinterface) from a virtual machine. The virtual machines are running on the same server.

There is no problem to access the server from another devices (e.g.:Notebook, Smartphone)

Server Setup:
- Server installed with Debian 11 and Cockpit.
- Enabled Services: fileserver and virtual machines.
- IP: 192.168.1.2
- Firewall disabled

Network:
- 192.168.1.0/24
- DHCP enabled 192.168.1.100-150

Virtual Machine:
- Linux and Windows
- Virtual Network interface: Direct attachment (Source: internal network card of the server)
- IP address DHCP -> 192.168.1.1xx
- Internet Access: Yes
- Access to Fileserver (192.168.1.2)?: No
- Can Ping Server (192.168.1.2)?: No
- Can open Cockpit Webinterface (192.168.168.1.2)?: No
- Can access, ping other devices in the network?: Yes

### Version of Cockpit

cockpit
276.1-1

### Where is the problem in Cockpit?

Virtual Machines

### Server operating system

Debian

### Server operating system version

11

### What browsers are you using?

Firefox, Edge, Chromium

**Comment by martinpitt** [comment]
Hello @err0r2k: To be honest I don't understand the description. Do you mean you created a VM using the "Machines" page on 192.168.1.2, and that VM cannot access other IPs on that network? Or that cockpit is running *inside* of that VM 192.168.100.120? That is a different network, so you'll need a route to the 192.168.1.0/24 network or it must be reachable through the default route.

Can you please describe the structure of the involved machines/VMs and what you did in a more detailed way? Copying & pasting exact commands, outputs, and information like `ip a` and `ip route` also helps.

**Comment by err0r2k** [comment]
@martinpitt 
I updated the description. I hope, the problem is now better described.
ip a and ip route will follow.

**Comment by err0r2k** [comment]
> Can you please describe the structure of the involved machines/VMs and what you did in a more detailed way? Copying & pasting exact commands, outputs, and information like `ip a` and `ip route` also helps.

ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s31f6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 90:1b:0e:f9:91:f4 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.2/24 brd 192.168.1.255 scope global dynamic noprefixroute enp0s31f6
       valid_lft 311000sec preferred_lft 311000sec
    inet6 fe80::7d7e:c27a:1ead:b67/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
20: virbr0: <BROADCAST,MULTICAST> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 52:54:00:c0:00:09 brd ff:ff:ff:ff:ff:ff

ip route
default via 192.168.1.1 dev enp0s31f6 proto dhcp src 192.168.1.2 metric 100 
192.168.1.0/24 dev enp0s31f6 proto kernel scope link src 192.168.1.2 metric 100

**Comment by martinpitt** [comment]
These `ip` commands were done from the server. These look alright, but your problem was that the VM running on the server cannot access the server? This is either a routing problem or the server is running a firewall.

Also, there is nothing cockpit specific here at all, this is just general networking.

**Comment by err0r2k** [comment]
> These `ip` commands were done from the server. These look alright, but your problem was that the VM running on the server cannot access the server? This is either a routing problem or the server is running a firewall.
> 
> Also, there is nothing cockpit specific here at all, this is just general networking.

That's correct.

Here is ip a and ip route form the virtual client.

linux@lmdevm:~$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp1s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:00:4a:f1:20 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.104/24 brd 192.168.1.255 scope global dynamic noprefixroute enp1s0
       valid_lft 604576sec preferred_lft 604576sec
    inet6 fe80::7571:2816:1861:d8e7/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever

default via 192.168.1.1 dev enp1s0 proto dhcp metric 100 
169.254.0.0/16 dev enp1s0 scope link metric 1000 
192.168.1.0/24 dev enp1s0 proto kernel scope link src 192.168.1.104 metric 100 

Yes, I cannot access the server with the VM. This is a simple network and firewall is not active. I can access the server with every other device. Only the VM's running on the server can't do it.

● firewalld.service - firewalld - dynamic firewall daemon
     Loaded: loaded (/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
       Docs: man:firewalld(1)

I tried to change the network card from the virtual client to "Bridge to LAN" (Source local network card from server) but I get the error message:
"Unable to add bridge enp0s31f6 port vnet0: Operation not supported".

**Comment by err0r2k** [comment]
Found the solution! I created a bridge in the "Networking" Tab in Cockpit. After this, it was possible to change the network card of the VM's to "Bridge to LAN". Everything is working now.

Moe information: With "direct attachment" it is not possible to communicate with the host.
Source:  [the network interface will not be be to communicate with other virtual machines running on the same host, nor with the host itself.](https://github.com/cockpit-project/cockpit-machines/issues/285)
